### PR TITLE
feat: safer rrweb events and regular full snapshots

### PR DESCRIPTION
**Pull request by pauldambra**
# regular full snapshots

we currently only take a full snapshot at the start of a recording or on return from idle

so if someone has an active and complex 20 minute recording we accrue 20 minutes of incremental snapshots. this has performance impact on playback and on seeking

instead now, when we start recording we set an interval to take another full snapshot in 5 minutes

if we see a fullsnapshot before that we reset the timer, so that we would take a full snapshot 5 minutes after that

# safer rrweb events

we believe we are missing full snapshots and custom events - maybe on return from idle

this adds a simple in-memory queue so that if we cannot take one of these triggered events, we will retry the next time rrweb emits an event

it also adds a custom event to report this happened so that we can track how often and by how much this is happening (cos it might not help 🤷)

### CODE REVIEW

File: src/extensions/replay/sessionrecording.ts [hunk 1]
```diff
@@ -464,26 +478,39 @@ export class SessionRecording {
...
         }
     }
 
     private _tryAddCustomEvent(tag: string, payload: any): boolean {
-        return this._tryRRwebMethod(() => this.rrwebRecord?.addCustomEvent(tag, payload))
+        return this._tryRRWebMethod({
+            // this should throw if rrwebRecord is not available
+            rrwebMethod: () => this.rrwebRecord!.addCustomEvent(tag, payload),
+            enqueuedAt: Date.now(),
+            attempt: 0,
```
Comment (marandaneto) [inline src/extensions/replay/sessionrecording.ts:503]:
```suggestion
            attempt: 1,
```

If it throws, it's technically 1 attempt?

Comment (marandaneto) [reply src/extensions/replay/sessionrecording.ts:503]: Same for _tryTakeFullSnapshot

File: src/extensions/replay/sessionrecording.ts [hunk 2]
```diff
@@ -584,6 +616,16 @@ export class SessionRecording {
...
     }
 
+    private _scheduleFullSnapshot(): void {
+        if (this._fullSnapshotTimer) {
+            clearInterval(this._fullSnapshotTimer)
+        }
+
+        this._fullSnapshotTimer = setInterval(() => {
+            this._tryTakeFullSnapshot()
+        }, 1000 * 60 * 5) // 5 minutes
```
Comment (marandaneto) [inline src/extensions/replay/sessionrecording.ts:626]: Move this to a constant, just a bit though, you spare the calculation on every method call unless the compiler already optimizes that.

File: src/extensions/replay/sessionrecording.ts [hunk 3]
```diff
@@ -654,6 +703,40 @@ export class SessionRecording {
...
+            // if they fail here, they'll be pushed into a new queue,
+            // and tried on the next loop.
+            // there is a risk of this queue growing in an uncontrolled manner,
+            // so its length is limited elsewhere
+            // for now this is to help us ensure we can capture events that happen
+            // and try to identify more about when it is failing
+            const itemsToProcess = [...this.queuedRRWebEvents]
+            this.queuedRRWebEvents = []
+            itemsToProcess.forEach((queuedRRWebEvent) => {
+                if (Date.now() - queuedRRWebEvent.enqueuedAt > 2000) {
```
Comment (marandaneto) [inline src/extensions/replay/sessionrecording.ts:721]: Name the magical numbers (eg a const) so its easier to understand if this is millis, seconds, etc.

**Review by marandaneto — APPROVED** [review]
Left a few comments but LGTM.
If we see that this improves something (eg player performance) or at least helps us debug things, I can do the very same on Mobile.

File: src/extensions/replay/sessionrecording.ts (hunk 2 above)
Comment (pauldambra) [reply src/extensions/replay/sessionrecording.ts:626]: I'd hope the compiler figures it out... but I already felt the need to add a comment... so a constant is less effort without having to guess what the compiler does 😊

Comment (pauldambra) [reply src/extensions/replay/sessionrecording.ts]: and I guess multiple browsers means multiple possible compilers so multiple behaviours

**Pull request merged**